tuf/tuf.go

```diff
@@ -210,32 +225,49 @@ func (tr *Repo) GetDelegationRole(name string) (data.DelegationRole, error) {
 	}
 	// Traverse target metadata, down to delegation itself
 	// Get all public keys for the base role from TUF metadata
-	signedTargetData, ok := tr.Targets[data.CanonicalTargetsRole]
+	_, ok = tr.Targets[data.CanonicalTargetsRole]
 	if !ok {
 		return data.DelegationRole{}, ErrNotLoaded{data.CanonicalTargetsRole}
 	}
 
 	// Start with top level roles in targets. Walk the chain of ancestors
 	// until finding the desired role, or we run out of targets files to search.
-	delegationRoles := signedTargetData.Signed.Delegations.Roles
-	var foundRole *data.Role
-	for len(delegationRoles) > 0 {
-		delgRole := delegationRoles[0]
-		delegationRoles = delegationRoles[1:]
-
-		// If we found the role, we can exit the loop
-		if delgRole.Name == name {
-			foundRole = delgRole
-			break
-		}
-
-		// If the current role is an ancestor of our desired role, add its children
-		// to the queue of roles to investigate.
-		if strings.HasPrefix(name, delgRole.Name+"/") {
-			if delegationMeta, ok := tr.Targets[delgRole.Name]; ok {
-				delegationRoles = append(delegationRoles, delegationMeta.Signed.Delegations.Roles...)
+	var foundRole *data.DelegationRole
+	buildDelegationRoleVisitor := func(tgt *data.SignedTargets, validRole data.DelegationRole) error {
+		if tgt == nil {
+			return ErrContinueWalk{}
+		}
+
+		// Try to find the delegation and build a DelegationRole structure
+		for _, role := range tgt.Signed.Delegations.Roles {
+			if role.Name == name {
+				pubKeys := make(map[string]data.PublicKey)
```

Comment:
Non-blocking - since this seems to be shared with the code in `SignedTargets.buildDelegationRoles`, would it make sense to pull this out into a helper function (at least the `BaseRole` part)?  It could also be used by `tuf.Repo.GetBaseRole`, etc.